src/components/data-table/ha-data-table.ts

```diff
@@ -89,95 +98,27 @@ export class HaDataTable extends BaseElement {
   @property({ type: String }) private _filter = "";
   @property({ type: String }) private _sortColumn?: string;
   @property({ type: String }) private _sortDirection: SortingDirection = null;
-
-  private _filterSortData = memoizeOne(
-    (
-      data: DataTabelRowData[],
-      columns: DataTabelColumnContainer,
-      filter: string,
-      direction: SortingDirection,
-      sortColumn?: string
-    ) =>
-      sortColumn
-        ? this._memSortData(
-            this._memFilterData(data, columns, filter),
-            columns,
-            direction,
-            sortColumn
-          )
-        : this._memFilterData(data, columns, filter)
-  );
-
-  private _memFilterData = memoizeOne(
-    (
-      data: DataTabelRowData[],
-      columns: DataTabelColumnContainer,
-      filter: string
-    ) => {
-      if (!filter) {
-        return data;
-      }
-      const ucFilter = filter.toUpperCase();
-      return data.filter((row) => {
-        return Object.entries(columns).some((columnEntry) => {
-          const [key, column] = columnEntry;
-          if (column.filterable) {
-            if (
-              (column.filterKey ? row[key][column.filterKey] : row[key])
-                .toUpperCase()
-                .includes(ucFilter)
-            ) {
-              return true;
-            }
-          }
-          return false;
-        });
-      });
-    }
+  @property({ type: Array }) private _filteredData: DataTabelRowData[] = [];
+  private _sortColumns: {
+    [key: string]: DataTabelSortColumnData;
+  } = {};
+  private curRequest = 0;
+  private _worker: any | undefined;
+
+  private _debounceSearch = debounce(
+    (ev) => {
+      this._filter = ev.detail.value;
+    },
+    200,
+    false
   );
 
-  private _memSortData = memoizeOne(
-    (
-      data: DataTabelRowData[],
-      columns: DataTabelColumnContainer,
-      direction: SortingDirection,
-      sortColumn: string
-    ) => {
-      const sorted = [...data];
-      const column = columns[sortColumn];
-      return sorted.sort((a, b) => {
-        let sort = 1;
-        if (direction === "desc") {
-          sort = -1;
-        }
-
-        let valA = column.filterKey
-          ? a[sortColumn][column.filterKey]
-          : a[sortColumn];
-
-        let valB = column.filterKey
-          ? b[sortColumn][column.filterKey]
-          : b[sortColumn];
-
-        if (typeof valA === "string") {
-          valA = valA.toUpperCase();
-        }
-        if (typeof valB === "string") {
-          valB = valB.toUpperCase();
-        }
-
-        if (valA < valB) {
-          return sort * -1;
-        }
-        if (valA > valB) {
-          return sort * 1;
-        }
-        return 0;
-      });
-    }
-  );
+  protected firstUpdated() {
+    super.firstUpdated();
+    this._worker = sortFilterWorker();
+  }
 
-  protected updated(properties: PropertyValues) {
+  protected async updated(properties: PropertyValues) {
```

Comment:
`updated` is a lifecycle method that we override, we cannot change the signature to be async.